src/modules/TasksModule/components/TaskesData/TaskesData.tsx

```diff
@@ -1,10 +1,220 @@
-
+import axios from "axios";
+import { SetStateAction, useEffect, useState } from "react";
+import { useNavigate ,Link} from "react-router-dom";
+import { toast } from "react-toastify";
+import { SubmitHandler, useForm } from "react-hook-form";
+import { baseUrl } from "../../../../Context/baseUrl";
+import Select from "react-select";
+interface FormValues {
+  title: string;
+  description: string;
+  employeeId?: number;
+  projectId?: number;
+}
+interface projectType {
+  id: number;
+  title: string;
+  description: string;
+}
 const TaskesData = () => {
+  const [userList, setUserList] = useState([]);
+  const [projectList, setProjectList] = useState([]);
+  const [selectedUser, setSelectedUser] = useState(null);
+  const [isLoading, setIsLoading] = useState(false);
+
+  const navigate = useNavigate();
+
+  const {
+    register,
+    handleSubmit,
+    formState: { errors },
+    setValue,
+    setError,
+  } = useForm<FormValues>();
+  const goBack = () => {
+    navigate(-1);
+  };
+  // Fucntion to add new tasks
+  const addNewTask: SubmitHandler<FormValues> = async (data: FormValues) => {
+    try {
+      const response = await axios.post(`${baseUrl}/Task`, data, {});
+      console.log(response);
+      toast.success("Task added suceessfully");
+      setIsLoading(false);
+      navigate("/dashboard/tasks");
+    } catch (error) {
+      toast.error(error.response.data.message);
+      console.log(error);
+    }
+  };
+  // fucntion to get the user's
+  const getUsersList = async () => {
+    try {
+      const response = await axios.get(
+        `${baseUrl}/Users/?pageSize=100&pageNumber=1`,
+        {}
+      );
+
+      setUserList(response?.data?.data);
+    } catch (error) {
+      toast.error("An error occurred. Please try again.");
+    }
+  };
+  // fucntion to get the Projects
+  const getAllProjectsList = () => {
+    axios
+      .get(`${baseUrl}/Project/?pageSize=10000&pageNumber=1`)
+      .then((response) => {
+        setProjectList(response?.data?.data);
+      })
+      .catch((error) => {
+        toast.error(
+          error.response.data.message || "An error occurred. Please try again."
+        );
+      });
+  };
+
+  useEffect(() => {
+    getUsersList();
+    getAllProjectsList();
+  }, []);
```

Comment:
It is recommended by React to include anything declared outside of the effect inside the dependency array.  So here we should put the function inside the dependency array, but we will encounter another issue, the function is created in every rendered and this will cause the effect to run in every render and might cause an infinite loop.
We can solve this by two ways:
1- by using `useCallback` hook to preserve the reference of the function between renders.
2- make another state variable responsible for re-fethcing the data as we did yesterday and declare the function inside the effect.
![image](https://github.com/RanaFathy123/pms-project/assets/64327685/800a1af2-2293-43b9-91f2-3964b30a201d)
 
You can read more about this here in the React docs:
https://react.dev/reference/react/useEffect#specifying-reactive-dependencies
https://react.dev/reference/react/useEffect#removing-unnecessary-function-dependencies.

Also, we can make one effect for each concern (one for the projects and one for the users). This will prevent calling APIs unnecessarily.